File: yggy/observable/messages.ts
import { Message } from "../comm/index.js";

export const OBSERVABLE_CHANGE_MSG = "observable.change";
export const OBSERVABLE_CLIENT_CHANGE_MSG = "observable.client_change";
export const OBSERVABLE_READY_MSG = "observable.ready";
export const OBSERVABLE_REGISTER_MSG = "observable.register";

export type ChangeMessage<T> = Message & {
    data_id: string;
    old_value: T;
    new_value: T;
};

export type ClientChangeMessage<T> = ChangeMessage<T> & {};

export type RegisterMessage<T extends "value" | "object"> = Message & {
    observable_type: T;
    data_id: string;
};

export type RegisterValueMessage<T> = RegisterMessage<"value"> & {
    value: T;
};

export type RegisterObjectMessage = RegisterMessage<"object"> & {
    attrs: { [key: string]: string };
};

export type ReadyMessage = Message & {
    client_id: string;
};

File: yggy/observable/messages.py
from ..comm import Message

OBSERVABLE_CHANGE_MSG = "observable.change"
OBSERVABLE_CLIENT_CHANGE_MSG = "observable.client_change"
OBSERVABLE_REGISTER_MSG = "observable.register"


class ChangeMessage[T](Message):
    data_id: str
    old_value: T
    new_value: T


class ClientChangeMessage[T](ChangeMessage[T]):
    ...


class RegisterMessage[T](Message):
    data_id: str
    value: T
